[mlir][llvmir] Added extra builders for CallInstrinsicOp

FMarno opened the pull request:
I've added the extra builders I would like to see for CallIntrinsicOp.
This is inspired by the comment from @antiagainst from [here](https://github.com/llvm/llvm-project/pull/108933#issuecomment-2392751569). I am also building Triton.

FMarno committed:
[mlir][llvmir] Added extra builders for CallInstrinsicOp

gysit commented on mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:
```diff
@@ -1943,6 +1943,11 @@ def LLVM_CallIntrinsicOp
   }];
 
+  let builders = [
+    OpBuilder<(ins "Type": $results, "StringAttr":$intrin, "ValueRange":$args)>,
```
```suggestion
    OpBuilder<(ins "Type": $result, "StringAttr":$intrin, "ValueRange":$args)>,
```
nit: maybe call it result or resultType since it is just one type.

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -3353,6 +3353,25 @@ struct LLVMOpAsmDialectInterface : public OpAsmDialectInterface {
+                            mlir::LLVM::FastmathFlagsAttr fastMathFlags) {
+
+  build(builder, state, results, intrin, args, fastMathFlags,
+        llvm::ArrayRef<mlir::ValueRange>{});
```
```suggestion
        /*op_bundle_operands=*/{});
```
nit: this is the operand bundle argument right? I think it is helpful to comment the default initialized arguments to illustrate what is going on.

gysit reviewed (commented):
Thanks it makes sense adding such builders!

Did you test it downstream? It is a bit hard to see from the code if all the necessary fields of the default constructor are set.

Lancern commented on mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:
```diff
@@ -1943,6 +1943,11 @@ def LLVM_CallIntrinsicOp
     attr-dict
   }];
 
+  let builders = [
+    OpBuilder<(ins "Type": $results, "StringAttr":$intrin, "ValueRange":$args)>,
+    OpBuilder<(ins "TypeRange": $results, "StringAttr":$intrin, "ValueRange":$args, "FastmathFlagsAttr":$fastMathFlags)>
+    ];
```
Maybe we could also include builders for intrinsics that do not have a result type?

```
OpBuilder<(ins "StringAttr":$intrin, "ValueRange":$args)>
OpBuilder<(ins "StringAttr":$intrin, "ValueRange":$args, "FastmathFlagsAttr":$fastMathFlags)>
```

FMarno replied:
Thanks for the review.
I've added those in. Do you know how I could confirm they work? I don't really have a use cause so I was just guessing.

Lancern replied:
Another approach might be adding a unit test under `mlir/unittests/Dialect/LLVMIR`. But it seems like few people ever do this ... Anyway this builder is fairly simple and I believe it should be OK if no reviewers have problems about it.

Lancern reviewed (commented):
Thanks for improving this!

FMarno committed:
fixup! [mlir][llvmir] Added extra builders for CallInstrinsicOp

FMarno committed:
fixup! [mlir][llvmir] Added extra builders for CallInstrinsicOp

FMarno commented:
@gysit Thank you for the review, I've addressed all your comments.

> Did you test it downstream? It is a bit hard to see from the code if all the necessary fields of the default constructor are set.

Yes, it works for code gen in Triton. If there is a better way to test builders then please let I'd love to know.

FMarno committed:
Added a couple more builders for call instrinsic op

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -3324,6 +3324,34 @@ LogicalResult CallIntrinsicOp::verify() {
 
+void CallIntrinsicOp::build(OpBuilder &builder, OperationState &state,
+                            mlir::StringAttr intrin, mlir::ValueRange args) {
+  build(builder, state, /*results=*/Type{}, intrin, args, FastmathFlagsAttr{},
```
```suggestion
  build(builder, state, /*results=*/TypeRange{}, intrin, args, FastmathFlagsAttr{},
```
I think this is a type range (probably Type converts to TypeRange)?

FMarno replied:
Thanks, I've addressed that.

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -3324,6 +3324,34 @@ LogicalResult CallIntrinsicOp::verify() {
+void CallIntrinsicOp::build(OpBuilder &builder, OperationState &state,
+                            mlir::StringAttr intrin, mlir::ValueRange args,
+                            mlir::LLVM::FastmathFlagsAttr fastMathFlags) {
+  build(builder, state, /*results=*/Type{}, intrin, args, fastMathFlags,
```
```suggestion
  build(builder, state, /*results=*/TypeRange{}, intrin, args, fastMathFlags,
```

gysit reviewed (approved):
LGTM modulo nits!

The only easy way to test builders is using them in a pass/canonicalization AFAIK. Since there are not many uses of this op upstream that does probably not work here.

FMarno committed:
fixup! Added a couple more builders for call instrinsic op
